proxy/http/HttpTransact.h

```diff
@@ -605,10 +605,9 @@ class HttpTransact
       }
 
       void
-      increment(MgmtInt configured_connect_attempts_max_retries)
+      increment()
       {
         ++_v;
-        ink_assert(_v <= configured_connect_attempts_max_retries);
```

Comment:
why wouldn't this assert apply now with the new changes? seems like increment doesn't have to change now that you check attempts - 1 before you increment

Reply:
This is a great question. The check before the increment ensures that `attempts - 1 < the global parent_max_retries`. The problem with this assert is that the argument passed in `configured_connect_attempts_max_retries` may not be the global parent max retry, but the max retry count for the current parent. Currently, the argument is set to the the output of `s->configured_connect_attempts_max_retries()`, which is returning:

- if connecting to the OS, returns the `txn_conf->connect_attempts_max_retries` 
- if connecting to a parent proxy, returns the max retries for the current parent(bounded by the global max retry for all parents)
In the second case, the assert can fail as after the increment, it went over the max for the current parent(which should be allowed as we switch to the next parent).

Reply:
but doesn't this line `auto cur_parent_max_attempts = ((cur_tries + ppca - 1) / ppca) * ppca;` take into account the max retries for the parent but also in relation to the total? even when we hit the max retries for the current parent, don't your new changes in `configured_connect_attempts_max_retries()` account for the next parent. each increment should be protected. im trying to find the edge case when this assert fails, could you provide an example

Reply:
The assert can be triggered if parent_connect_attempts > (connect_attempts_max_retries + 1).  If you can find checking logic in the config code that enforces this, it would be safe to restore the assert.

This was the original version of increment():
```
      void
      increment()
      {
        ++_v;
        ink_assert(_v <= configured_connect_attempts_max_retries());
      }
```
This was CurrentInfo::configured_connect_attempts_max_retries():
```
inline MgmtInt
HttpTransact::CurrentInfo::Attempts::configured_connect_attempts_max_retries() const
{
  return reinterpret_cast<const HttpTransact::State *>(reinterpret_cast<const char *>(this) -
                                                       offsetof(HttpTransact::State, current.attempts))
    ->configured_connect_attempts_max_retries();
}
```
Even though this code was verified in Yahoo prod, and this sort of pointer arithmetic is how inheritance is implemented in C++, this code was rejected in review.  I never would have put in the assert if I thought in would require adding a parameter to increment().

Reply:
> even when we hit the max retries for the current parent, don't your new changes in `configured_connect_attempts_max_retries()` account for the next parent?

`configured_connect_attempts_max_retries()` normally would be increment to the next multiple of `ppca`(accounts for the next parent), except for the case where the attempt/(retry + 1) is already a multiple of ppca(current parent hits max retries), in which case the calculation basically returns the same attempt/retry count as before. 

> im trying to find the edge case when this assert fails, could you provide an example

The above logic make sense when we call `maximize()` with the `configured_connect_attempts_max_retries()` to disable further retries, since we already hit the max retries for the current parent, we are not setting it to anything larger.

However, it will fail the assert in `increment()` with the `configured_connect_attempts_max_retries()`, since `n = configured_connect_attempts_max_retries(n).`